### ISSUE TITLE
Write tutorial

### ISSUE THREAD

**Issue by SimonGreenhill** [lmaurits/BEASTling#59]
...include example data.

**Comment by lmaurits** [comment]
+1

**Comment by Anaphory** [comment]
What's a good format for a tutorial, and how do we make sure it ties in well with tests and documentation?

**Comment by lmaurits** [comment]
I  have been envisaging one or two tutorial datafiles to be included in the repository, probably one lexical and one typological (no serious ideas on suitable candidates yet, except Mattis can probably provide a good lexical file, but other suggestions welcome).  The tutorial will basically consist of example BEASTling config files for each kind of analysis, ratcheting up complexity (e.g. starting with strict clock and no rate variation, then adding those one at a time, etc.).

Re tying in with documentation: to my mind, the tutorial _is_ documentation.  It can and probably should link to relevant sections of the fuller documentation.  Re: tying in with tests, I suppose we should have a set of tests making sure that all of the example analyses run, I'm not sure anything is required beyond that?

Doing a really good job of this is probably going to take much longer than the envisaged timeline for the 1.2.0 release.  I'm happy for the version included with that to be "good" as opposed to "great".

**Comment by Anaphory** [comment]
I was thinking along lines such as the following:

There are some ways to turn examples from docstrings into test-assertions, which would tie function documentation into tests. I wonder whether there is a similar way for separate files, which would make sure that the tutorial does not become obsolete because we forget to update it. (The tests being pulled directly from the tutorial would mean that that would throw test failures right and left.)

For tying in with the documentation, of course tutorial is part of the documentation. I was thinking along the lines of pulling the necessary explanation snippets directly from somewhere eg. in the docstrings, so the tutorial would also be a test of the documentation.

I also asked about the format, because I was thinking along the lines of some python library tutorials that are presented in a jupyter ipython notebook or something similar interactive. Of course, that applies less to a command line tool (and the command line tool-ness may also make my first point more difficult), but I thought maybe someone has an idea like that.

**Comment by Anaphory** [comment]
The doctest module (https://docs.python.org/3/library/doctest.html) can work on rst documentation text files. That's still not helping with the CLI step, but there may be ways around that.

I guess for the moment, I will use a similar syntax to doctest: I shall make sure my examples start indented, and the first line of command line input will be `$`. Subsequent lines of command line input I will start with a different symbol, I guess `>`. (I hope this won't lead to issues with being misinterpreted as “quotation”.) Output will be just indented, and three dots (`[...]`) will show that the output was shortened. Square brackets `[]` may contain other comments on the output, so `[... Same output as above ...]` should be valid bash-doctest.

I can either supply files in shell syntax (`$echo >`) or make up my own syntax, what do you consider better?

**Comment by Anaphory** [comment]
Should the tutorial actually be similar/growing into the analysis we use in the paper, or be something entirely different?

**Comment by Anaphory** [comment]
How do I want to specify files in the bash-doctest tutorial examples? I can see four ways, none of them entirely convinces me (I think I like the first one best, but I haven't tested how Sphinx actually represents it):
1. Use an attributed line block quote or an attributed source-code block quote:
   
   ```
   |[mcmc]
   |chainlength=10
   --- example.conf
   ```
   
   or
   
   ```
   ::
       [mcmc]
       [chainlength=10
   --- example conf
   ```
2. Use bash-doctest with nearly-copyable echo:
   
   ```
   $ echo <<DELIMITER
   > [mcmc]
   > chainlength=10
   > DELIMITER > example.conf
   ```
3. Use cross-platform echo, which needs significant effort to copy&paste into a text editor
   
   ```
   $ echo '[mcmc]' > example.conf
   $ echo 'chainlength=10' >> example.conf
   ```
4. Break the bash-doctest assumption and make `cat` special so that it does not match expected output, but writes the “expected” output to file.
   
   ```
   $ cat example.conf
   [mcmc]
   chainlength=10
   ```

This is purely for files that the user is supposed to create. File inspection, eg. from downloads, should use the usual

```
$ cat austronesian.csv
Language_ID,Feature_ID,Value
aaaa1111,1,1
[...]
```

**Comment by Anaphory** [comment]
Normally, I would want the test to die when one of the commands fails, but on the other hand I do want to introduce `--overwrite` by having a beastling fail after the output was not cleaned up – and in general, it may be useful to show failure modes and how to deal with them.

Shall I just continue on the assumption that failure produces output which doesn't match the expected output?

**Comment by Anaphory** [comment]
It looks like `scriptdoctest` might be writable – but does someone have a good intuition what to do about a script's `stdout` vs. `stderr`?
- If I imagine a shell example, I cannot easily distinguish between what was output to `stderr` and `stdout`. So I would expect that the “expected output” in an example should also look the same.
- Using `subprocess`, output is strongly separated into `stdout` and `stderr`, which makes it hard to match them to a merged output, although we could force a redirect (`2>`) in our call to Popen.
  - Eclipse distinguishes `stdout` and `stderr` by colour. This is not a bad option, but it means that examples are not a simple copy&paste from the command line, because they need to be annotated for `stdin` vs `stderr` using some ReST directives (or magic symbols, but the magic symbols would then show up in the docs).

**Comment by Anaphory** [comment]
It's now a thing:

I spent today implementing some core functionality for [scriptdoctest](https://github.com/Anaphory/scriptdoctest/). It doesn't work as command line tool yet, supplies no documentation, has no syntax for existing files, needs a slightly patched `scripttest`, duplicates many private methods from `doctest`, among other flaws. But I will try to see how it deals with the tutorial, and so can you.

**Comment by lmaurits** [comment]
@SimonGreenhill Is it okay to be using Lexibank data for the examples [here](https://github.com/lmaurits/BEASTling/blob/develop/docs/tutorial.rst#using-beastling), prior to official Lexibank launch?

**Comment by lmaurits** [comment]
I have pushed a little bit of extra text, not as much as I'd have liked but I'll continue to work on this for the rest of the week.  I've not yet taken the time to familiarise myself with how `scriptdoctest` works so I've not put an awful lot of shell content in there.  Random thoughts:
- To some extent the stuff on installing dependencies duplicates existing documentation in `docs/installation.rst`
- I wonder if we should use covarion for the cognate example.  I know the surrounding wordage emphasises that this is a naive and untrusted analysis, but all the same, maybe it's best to make the tutorial resilient against blind copy and pasting.  And, no, I don't really like the binary covarion model either, but it's the most cognatey thing that works ATM.
- Were we thinking WALS for the typological data?  1.2.0 will be released well, well before Grambank.
- I assume there is a way to embed images in .rst files and have GitHub render them?  This will be good for including screenshots of things like Tracer, FigTree, etc.
- Whatever data we end up using, should we put the relevant files in the repo, rather than asking the tutorial followers to download them from possibly changeable/breakable third party links?

**Comment by xrotwang** [comment]
Currently, the lexibank repos doesn't contain any CLDF data - just code to
create it. So the tutorial would have to describe that workflow as well -
at least for the time being.
Am 12.10.2016 04:56 schrieb "Luke Maurits" notifications@github.com:

**Comment by Anaphory** [comment]
- `scriptdoctest` code examples just look like
  
  ```
  
     $ echo "Two lines
     > of output"
     Two lines
     of output
  
  ```
  
  at the moment. There will be a way to specify files added today. Input is run, and compiled to output literally. I still need to figure out what to do with `...` etc, for the moment I'm using `doctest`'s output matcher, so the `doctest` documentation applies.
- I know the stuff on installing dependencies might duplicate parts of `installation.rst`.
  I think the tutorial is the point to explain how to install BEAST, Python and maybe Java, and then should just assume `pip` exists and use that.
  
  The `installation.rst`, I thought, would _mention_ you need Python and, usually, BEAST and therefore Java 8. On the other hand, it would describe all dependencies of BEASTLING, and mention `github` and `setup.py` and so on, which the tutorial won't.
- Yes, the tutorial should demonstrate best practices in what model to use. If that means for now using `covarion`, that's what it should do. It also means sampling from prior and comparing, and showing at least one cross-validation.
- When I looked at possible data, I thought we might also point out something like [SAILS](http://sails.clld.org), but then we would have to get some South American vocabulary data – Arawakan is available, for example, but only in PDF (!) format, Uto-Aztecan should live somewhere, but I could not find it properly, etc.
- I'm sure there is a way to embed images. There is an example in one of the “Documentation with ReST and Sphinx” tutorials.
- I would put them in the repo for testing purposes and still show the download. This means our tests will break when the link changes, and the user sees how easy it is to use data from the web.
- I tried and seem to have failed downloading `abvd` from lexibank-data. I should probably make a lexibank-data issue out of that.

**Comment by Anaphory** [comment]
There is now a [`scriptdoctest` example file.](https://github.com/Anaphory/scriptdoctest/blob/master/examples.py)

Note that a code block needs to be introduced by `::\n\n` for sphinx and scriptdoctest to recognize it. For files, the `::\n\n` and the `---` need to be on the same indentation level, everything in between needs to add indentation.

**Comment by lmaurits** [comment]
I think the biggest hold up right now to really making rapid progress on this is not having settled on a dataset, which makes it impossible for us to actually run the tutorial analyses, so we should sort this out as a priority.

Asking tutorial participants to programmatically generate CLDF data using Lexibank code is a bad idea, the barrier to entry for the tutorial should be as low as possible (and it's already higher than many would like, with regard to having to possibly install Python and Java and dependencies, and then get familiar with the CLI).  Ideally we should be able to host the data ourselves.  It doesn't need to be very large or even terribly high quality, as long as it's something that will be recognisably similar for most linguists to data they have themselves.

I just checked out the ABVD webpage, but it seems like there isn't a direct download available from there, and also a less-than-welcoming notice about not doing anything with the data without permission (and also a notice forbidding commercial use which is shortly thereafter flatly contradicted by a CC-BY license as opposed CC-BY-NC), which means I can't start working with it literally today, which is what I'm looking for right now.

In fact, having direct downloads of data in a nice simple .csv format is apparently very unfashionable, I could not find this for IELex or Uralex either.

I am tempted to go with a CLDF-ified version of the [Indo-European data](https://github.com/glottobank/BEASTling_paper/blob/master/examples/indoeuropean/PIE.csv) we are using as an example in the paper, simply because of immediate availability and the very high probability of Mattis being okay with us using it here (since he has already okayed using it in the paper).  Perhaps this makes sense anyway, people can read the paper and then see that example analysis built up from simpler ones.

I don't think it's necessary to insist on the typological and lexical examples being from the same language families, although it would be kind of nice.  SAILS is a possibility, but I've tried to build trees with it before, and there is just no signal in there at all, at least not with simple modelling.  The appeal of WALS is that it is right there, it's CC licensed so we can just use it, and it's also global which gives us a good opportunity to demonstrate using `families=` and/or `macroareas=` to pull out a subset of the data.

**Comment by lmaurits** [comment]
In a comment on a pull-request, @Anaphory mentioned the idea of having a dedicated branch of the repo for working on the tutorial.  I think this idea has promise (or rather, a general documentation branch).  Right now the tutorial is the only remaining issue tagged for the 1.2.0 milestone.  If we insist on treating documentation like code with clearly defined "releases" then we can't cut the 1.2.0 release until the tutorial is complete, which is not likely to happen this week.  But this leaves us in the unfortunate situation that, say, @xrotwang can't easily work on the paper reproducibility environment next week in parallel with @Anaphory and I working on the tutorial, because `pip` won't be able to install the latest BEASTling.  It also means that whenever we find an error or incompleteness in the tutorial, we're forced to do a new release, even if no changes have happened to the code, which seems a pain.  I feel like treating documentation as a separate conceptual entity could be advantageous.  What do people think?

**Comment by xrotwang** [comment]
What about a 1.2a1 release candidate?

**Comment by lmaurits** [comment]
I wasn't tremendously keen on this at first, but I see now that PyPi actually supports these quite nicely, in that `pip` will ignore prereleases by default so ordinary users (which admittedly are purely hypothetical entities for BEASTling at this point) will not end up upgrading before we are ready, but they can be explicitly requested in, say, a `requirements.txt`, so this seems like the ideal solution to our problem.

I only wish that the naming scheme to do this was `-rc1` instead of `.rc1` so it was fully compatible with the Semantic Versioning standard, but oh well.

I have quite a bit of work to do on the Eurasian project today, but I'm really hoping I'll be able to put 1.2.0.rc1 up on PyPi this afternoon or this evening (which will be tomorrow morning for you folk, so that should work out nicely).  Hopefully it won't take long, 90% of the time will be staring at Git documentation and trying to convince myself that I'm not about to horribly maim the repo with a new release branch and merge dance.  I think we should follow the established (problematic) Git scheme for 1.2.0 and then refactor for simplicity after that.

**Comment by Anaphory** [comment]
Would it not make life easier now and in the near future to throw away some of the recent erroneous commits on `master` and go from there, instead of adding on top of those in an undisentangleable mess?

**Comment by lmaurits** [comment]
There should now exist a `1.2.0rc1` distribution on PyPi, accessible via `pip` and  suitable for developing the reproducability environment in the paper repo.

**Comment by lmaurits** [comment]
> Would it not make life easier now and in the near future to throw away some of the recent erroneous commits on master and go from there, instead of adding on top of those in an undisentangleable mess?

Maybe?  I will begrudgingly read up on git voodoo later today and decide if that's true.  I know there is a big stigma against "rewriting history" in Git-land, but at this point I'm happy to be pragmatic about it to make 1.2.0 happen smoothly and quickly.  There's some quote from Torvalds somewhere about how he once merged 12 patches to the same kernel file at the same time, blindfolded, in the snow, and git just figured it all out and made it work, but in my experience it will fall over if you look at it funny.

At any rate, prior to merging the new `release-1.2` branch with `master` for the release, I guess the thing to do now is to continue working on the tutorial in the `release-1.2` branch and pull those changes back into `develop` when we're done.  I'm going to work on that for the rest of today (at last!).

**Comment by Anaphory** [comment]
On 19 October 2016 at 04:09, Luke Maurits notifications@github.com wrote:

> > Would it not make life easier now and in the near future to throw away some of the recent erroneous commits on master and go from there, instead of adding on top of those in an undisentangleable mess?
> 
> Maybe? I will begrudgingly read up on git voodoo later today and decide if that's true. I know there is a big stigma against "rewriting history" in Git-land, but at this point I'm happy to be pragmatic about it to make 1.2.0 happen smoothly and quickly.

I know someone who describes git as a “basic history-rewriting tool”.
But actually, I did not mean much history-rewriting. I meant mostly
renaming a branch, and doing some merging:

release-1.0 has not been merged into develop, which is a thing that
should happen anyway, whatever scheme we intend to follow. (This might
just be a formal “merge take ours”, so things will be trivially ahead
of those changes? – if all changes are superseded, this may not be
necessary, but I don't know git internals well enough, either.)

After that, https://github.com/lmaurits/BEASTling/commit/1660c5e59515a02a7bdc65833197625954aa1663
is the latest release, so should be the latest commit on `master`, the
pull request from Simon should be ignored/on a different branch/thrown
out and forgotten when no branches point to it.

Now you _could_ do history-rewriting around that commit, but that's a
hassle and those commits were _released_, so leave it at that.

That leaves you with a `master`, which is ahead of
https://github.com/lmaurits/BEASTling/commit/31ea30028291479dd6b974e5cfc6c9994d9b0d13
by some 4 commits*, and a develop which is ahead of `that particular
commit` by many commits.

*) Effectively, adding requirement versions and bumping version
numbers – If you don't merge release-1.0 first, those commits will
come on top, and will lead to conflicts down the line time and again,
I assume.

**Comment by lmaurits** [comment]
@Anaphory, is there any chance you can give the tutorial a little attention with respect to `scriptdoctest` compatibility?  Regrettably I've still not had the chance to really play with it yet.

**Comment by Anaphory** [comment]
I should be able to do so soon, I wanted to read it anyway beyond the nice “MCMC Robot” thing, which I noticed in a commit earlier.

**Comment by lmaurits** [comment]
Glad you liked that :)

**Comment by lmaurits** [comment]
I have just made what are hopefully the final changes to the tutorial text, with the possible exception of minor tweaks and corrections.  I am planning to release 1.2.0 this week come hell or high water.  If you can get the `scriptdoctest` stuff looking and working nicely in that timeframe, it would be much appreciated.  If not, I will temporarily pull it out for the release and it can go back in afterward and hopefully be ready for the next one.

**Comment by Anaphory** [comment]
I have battled my university-provided desktop into a state that is reasonable to work with, and this is the next thing for me to do now.

**Comment by Anaphory** [comment]
I cannot get the abvd from here:

     $ curl -OL https://lexibank.clld.org/contributions/abvd.csv

**Comment by xrotwang** [comment]
No. Since lexibank isn't public yet, using it for a tutorial now is
probably not a good idea anyway.
Am 22.11.2016 15:24 schrieb "Gereon Kaiping" notifications@github.com:

**Comment by Anaphory** [comment]
Ah, sorry. That came from a missing pull (or rather, pulling the wrong branch). I should say, though, that the following happens:

    $ curl https://raw.githubusercontent.com/lmaurits/BEASTling/master/docs/tutorial_data/ie_cognates.csv
    404: Not Found

Which is not good either.
I have taken the liberty to specify the `release-1.2` branch in that url instead.

The next issue I'm now investigating is why scriptdoctest refuses to create the configuration file. It is unfortunate that the preformatted-text-in-an-attributed-blockquote structure is apparently slightly fragile.

**Comment by Anaphory** [comment]
Aha! It's not `scriptdoctest` that fails me, it's `scriptdoctest` doing its job!

The filenames `ie_cognates.conf` and `ie_vocabulary.conf` are different, so it *should* fail!

**Comment by lmaurits** [comment]
I have accepted your pull request and used Sphinx to rebuild the tutorial.  Is the fact that I can see things like `– ie_vocabulary.conf` underneath the contents of files by (grudging) design?  i.e. it's not so much that you want it there, but that you have to signal to scriptdoctest somehow that this is to be checked against file contents, and RST doesn't have any way of inserting this metadata invisibly, and this is the least offensive looking solution you've found?

**Comment by Anaphory** [comment]
Huh? Didn't I already post this answer?

I think it's good to keep filenames with example files, in particular in a tutorial like this. That doesn't mean that the current representation in ReST logic I chose (“attributed blockquote containing pre-formatted text”) is the best way to visually represent that, although I definitely couldn't find a better way in ReST semantics.

In a perfect world, we'd get a new ReST (or whatever) directive that stands for “File content” and will be translated by Sphinx into the file name in small text, which is a download link to the content of the file, followed by a syntax-highlighted, maybe line-numbered, copy-able version of the file. Given that this is not a perfect world, I'm actually quite okay with the current way to describe file contents and have a far bigger problem with many other things in my `scriptdoctest` implementation, but those problems are out of scope here.

**Comment by lmaurits** [comment]
Sorry, you're right, I read your earlier posts about the concept of `scriptdoctest` more carefully and you did explain the situation with this.  I have made a few tweaks (mostly for visual consistency of stuff which is supposed to be computer output) and you can see the results as rendered by Sphinx at https://beastling.readthedocs.io/en/release-1.2/tutorial.html  It is by no means perfect, but I consider it good enough for a first effort.

Please verify that my tweaks have not broken `scriptdoctest`.  If not, I will at last close this issue.

**Comment by Anaphory** [comment]
I did not mean “You can read my answer above”, I meant “I thought I had already written a comment like this and submitted it, where has it gone?”

I will check it out. I have not seen every example part succeed yet, because (a) waiting for Beast output to be produced is slow and then matching it is a pain, and (b) an Emacs bug made it hard to work on some part of the tutorial (the `head beastling.log` output, to be precise), due to repeated segfaults.

**Comment by lmaurits** [comment]
I understand from your PR commentary that `scriptdoctest` is now both running and passing on the tutorial as it stands.  While I definitely want to improve it in the future, I think the tutorial as it stands does a good enough job of covering the basics that this issue can be closed.  Further refinements can be discussed in dedicated issues (like the few @Anaphory has already opened).

I close this issue, and in so doing Milestone 1.2.0 is reached!  Thanks all for your hard work.

**Comment by Anaphory** [comment]
I should really get `scriptdoctest` onto pypi, so we can properly integrate it into the CI. (Although that will mean intalling and running Beast, maybe including packages, for every Master-PR. Augh. Do we want that?)
